openvdb/math/Stencils.h

```diff
@@ -1701,6 +1705,177 @@ class DenseStencil: public BaseStencil<DenseStencil<GridT, IsSafe>, GridT, IsSaf
+
+    void init(const Coord& ijk)
+    {
+        const Coord shape(
+            static_cast<Int32>(mKernelShape[0]),
+            static_cast<Int32>(mKernelShape[1]),
+            static_cast<Int32>(mKernelShape[2]));
+        const Coord offset = shape >> 1;
+
+        if (mInitialized && (ijk == this->getCenterCoord().offsetBy(0, 0, 1))) {
+            // If the stencil buffer has been initialized and its new position differs
+            // from the old by +1 along the z axis (the common case for convolution),
+            // preserve the contents of the buffer and sample only the new z slice.
+            // (Note that BaseStencil::moveTo() sets mCenter before calling init(),
+            // inhibiting this optimization.)
+            /// @todo This optimization could be applied for any offset that leaves some
+            /// overlap between the old and new stencils, but the bookkeeping gets messy.
+
+            typename BaseType::BufferType newStencil(mStencil.size());
+            const StencilRefType oldStencil(mStencil.data(), mKernelShape);
+
+            int n = 0;
+            Coord p = ijk - offset;
+            p[2] += shape[2] - 1;
+            for (Int32 i = 0; i < shape[0]; ++i, ++p[0]) {
+                p[1] = ijk[1] - offset[1];
+                for (Int32 j = 0; j < shape[1]; ++j, ++p[1]) {
+                    // Copy values from the old stencil to the new stencil,
+                    // shifting them by one voxel.
+                    for (Int32 k = 1; k < shape[2]; ++k) {
+                        newStencil[n++] = oldStencil[i][j][k];
+                    }
+                    // Sample one new voxel value from the grid.
+                    newStencil[n++] = mCache.getValue(p);
+                }
+            }
+
+            mStencil.swap(newStencil);
+
+        } else {
+            // Populate the stencil by sampling the grid at every voxel.
+            int n = 0;
+            Coord p = ijk - offset;
+            for (Int32 i = 0; i < shape[0]; ++i, ++p[0]) {
+                p[1] = ijk[1] - offset[1];
+                for (Int32 j = 0; j < shape[1]; ++j, ++p[1]) {
+                    p[2] = ijk[2] - offset[2];
+                    for (Int32 k = 0; k < shape[2]; ++k, ++p[2]) {
+                        mStencil[n++] = mCache.getValue(p);
+                    }
+                }
+            }
+        }
+
+        this->setCenter(ijk);
+    }
+
+    template<typename, typename, bool> friend class BaseStencil; // allow base class to call init()
+    using BaseType::mCache;
+    using BaseType::mStencil;
+    const KernelType mKernel;
+    const std::vector<size_t> mKernelShape;
+    bool mInitialized = false;
+}; // class ConvolutionStencil
+
+
+////////////////////////////////////////
+
+
+/// Return a 3D sharpening kernel of the given radius, for use with ConvolutionStencil.
+template<typename ValueType>
+ConvolutionKernel<ValueType>
+sharpeningKernel(int r)
```

Comment:
I'm not thrilled about moving `sharpeningKernel` here, but it's needed by both `tools::Filter` and `tools::LevelSetFilter`, and I don't see a better place to put it, other than in a header of its own.